### PR TITLE
Allow teams to sign up for Usage-Based Billing via Stripe

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

Allow teams to sign up for Usage-Based Billing via Stripe.

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/10326

## How to test

1. Join the "Gitpod" team: [invite link](https://jx-stripe-service.preview.gitpod-dev.com/teams/join?inviteId=9d9cd3f1-bad9-4725-8e06-469ce0d955c1) (this enables the Usage-Based feature flag for you)
2. Create a new, different team ✨ 
3. Navigate to your Team Settings > Team Billing
4. Wait for the Usage-Based UI to appear above the Team Plans
5. Upgrade your team to usage-based by providing a fake credit card (e.g. 4242 4242 4242 4242, 4/24, 424)
6. This should successfully create a Stripe Customer for your team, visible in our [Stripe Dashboard](https://dashboard.stripe.com/test/customers) (credentials in 1Password)

| No Team Billing | Add Credit Card | Active Team Billing | Stripe Dashboard |
| --- | --- | --- | --- |
| Screenshot 2022-06-01 at 16 54 30 | Screenshot 2022-06-01 at 16 55 02 | Screenshot 2022-06-01 at 16 58 35 | Screenshot 2022-06-01 at 16 56 42 |

## Release Notes

```release-note
NONE
```

## Documentation

- [x] /werft with-payment=true

### CODE REVIEW

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1837]: When interactions with 3rd party services fail, we shouldn't be propagating the errors directly to the client. We don't entirely control what is in the errors and may contain sensitive data. How is that handled here? I'd expect that we check for success/error and return appropriate obfuscated error messages upstream to the client.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1832]: This method should also be behind a feature flag. The feature flag on the client is useless if we do not check it on the server side too.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1837]: Agree -- we could avoid propagating to the user/client with a try/catch:

```ts
// Pseudo-code
try {
    await stripeThing();
} catch (error) {
    log.error("the full error for our logs", { error });
    throw new Error("Stripe thing failed for some reason ¯\_(ツ)_/¯");
}
```

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1832]: Thanks, but in this case I would disagree:
- Gating this behind the Usage-Based Feature flag would be over-engineered here (requires making yet another call to configcat from the server, and wait for the response which can take a while)
- We primarily want to hide Usage-Based Billing from users because it's work-in-progress, but if someone is dedicated enough to call the server API directly, I guess they're allowed to try out our Stripe integration (which we're planning to GA in a few weeks anyway)

**Comment by easyCZ** [comment]
Behaviour wise: works well and tested.

The popup to add credit-card logs a number of errors when making a `POST https://play.google.com/log?format=json&hasfast=true&authuser=0 net::ERR_BLOCKED_BY_CLIENT`. Can this be removed? Why does it happen?

If you wanted to, this PR could've been split into the following:
1. UI to show upgrade button which does nothing
2. Server-side APIs (behind a feature flag) which return Unimplemented when part of team gitpod, NotAllowed otherwise.
3. Server-side APIs actually implement behaviour
4. UIs which hook up the behaviour.

That way, you'd be able to keep the PRs smaller, and individual steps smaller. For example, in this case you'd get the review feedback on needing the feature flag guard server-side on the API setup, not on the full PR.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1832]: > We primarily want to hide Usage-Based Billing from users because it's work-in-progress, but if someone is dedicated enough to call the server API directly, I guess they're allowed to try out our Stripe integration (which we're planning to GA in a few weeks anyway)

The client must not be authoritative. We are not just hiding UI features, we're gatting functinality to control how we roll it out. We don't want anyone without the correct permissions (or membership) to access this feature - this cannot be done on the UI and must be done on the server side.

> making yet another call to configcat from the server, and wait for the response which can take a while)

Feature flags are updated async, actual eval is non-blocking.

Comment (geropl) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1832]: @jankeromnes I agree with @easyCZ : Deciding whether or not to enable a feature is a authorative (server-side) question. IMO it's the only way to serve this as "clean cut".
And whether or not we do another request against configcat should not guide any decision I think. :+1:

Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1833]: Done ✅

File: components/dashboard/src/teams/TeamUsageBasedBilling.tsx
Comment (jankeromnes) [reply components/dashboard/src/teams/TeamUsageBasedBilling.tsx]: Thanks! Fix implemented. ✅

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1833]: Should also be added for the other API methods that are for usage based billing.

**Comment by jankeromnes** [comment]
@gtsiolis Had a quick look into how we can style Stripe's PaymentElement.

[Here](https://stripe.com/docs/elements/appearance-api) is how they suggest we do it:

> 1. Start by picking a [theme](https://stripe.com/docs/elements/appearance-api#theme)

The 4 themes are:

| stripe | night | flat | none |
| --- | --- | --- | --- |
| Screenshot 2022-06-07 at 15 10 36 | Screenshot 2022-06-07 at 15 07 12 | Screenshot 2022-06-07 at 15 11 00 | Screenshot 2022-06-07 at 15 11 34 |

I think a good Minimum-Viable-Change could be:
- Use `stripe` or `flat` in Light theme
- Use `night` in Dark theme

> 2. Customize the theme using [variables](https://stripe.com/docs/elements/appearance-api#variables)
> 
> Set variables like `fontFamily` and `colorPrimary` to broadly customize components appearing throughout each Element.

We could do this to further align the look & feel of Stripe Elements with our dashboard in broad strokes. 💡

> 3. If needed, fine-tune individual components and states using [rules](https://stripe.com/docs/elements/appearance-api#rules)

With this, we can even drill down into specific styling issues that might remain after 1. and 2. 🔍

File: components/dashboard/src/teams/TeamUsageBasedBilling.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamUsageBasedBilling.tsx:125]: **nitpick**: Minor spacing fix to match the design specs.

```suggestion
                <div className="mt-4 h-32 p-4 pt-4 pl-4 flex flex-col rounded-xl bg-gray-100 dark:bg-gray-800">
```

File: components/dashboard/src/teams/TeamUsageBasedBilling.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamUsageBasedBilling.tsx:174]: **issue**: Although flat looks slightly better, should we default to the `stripe` theme so that UX looks similar with the rest of the product? 

| Variable Modal | Billing Modal |
|-|-|
| Screenshot 2022-06-07 at 5 39 49 PM (2) | Screenshot 2022-06-07 at 5 39 40 PM (2) |

**Review by gtsiolis — COMMENTED** [review]
@jankeromnes I left some minor and non-blocking comments regarding UX. 🏓 

Regarding the comments in https://github.com/gitpod-io/gitpod/pull/10378#issuecomment-1148673742:

1. Going with stripe's components sounds fine. Could we try updating the element styles to match our own? Do you prefer to skip it for now?
2. Did you figure out whether we can drop the country selection in the billing modal?

Also, thanks for taking a closer look on how we can customize Stripe's elements and including all the helpful references above in https://github.com/gitpod-io/gitpod/pull/10378#issuecomment-1148673742. 🙏

File: components/dashboard/src/settings/Billing.tsx
Comment (jankeromnes) [reply components/dashboard/src/settings/Billing.tsx:25]: Thought that could be helpful to Gitpodders with the feature-flag enabled. 💭

(Also the weird code artifacts are mandated by our auto-formatter 🙄.)

**Comment by jankeromnes** [comment]
> @jankeromnes I left some minor and non-blocking comments regarding UX. 🏓

Many thanks @gtsiolis! ❤️ Happy to opportunistically get them in when possible.

> 1. Going with stripe's components sounds fine. Could we try updating the element styles to match our own? Do you prefer to skip it for now?

Cool! And yes we could try that, but in a follow-up Pull Request. I think the initial Light/Dark switch I implemented is good enough for now, and we can easily extend it later with better custom styles/rules. 🎨

> 2. Did you figure out whether we can drop the country selection in the billing modal?

Yes, I believe it's documented somewhere here: https://stripe.com/docs/js/elements_object/create_payment_element

But, I'd love to leave this one as a follow-up too.

File: components/dashboard/src/teams/TeamUsageBasedBilling.tsx
Comment (gtsiolis) [reply components/dashboard/src/teams/TeamUsageBasedBilling.tsx]: Ah, you're right. I knew I should be more careful with sharing sandbox design files and not final design specs that include only designs for one issue. Sorry for the confusion, @jankeromnes!

You probably checked out a previous iteration[[1](https://www.figma.com/file/8UO0eql1UPmI1B0zgf5Kcu/Plans-(Sandbox)?node-id=1021%3A3469)], not the latest in progress below[[2](https://www.figma.com/file/8UO0eql1UPmI1B0zgf5Kcu/Plans-(Sandbox)?node-id=1030%3A3958)].

| Early iteration | Last iteration |
|-|-|
| Team Plans | Team Plans-1 |

**Comment by jldec** [comment]
Works for me as well - Is there a way to navigate manually to the Stripe portal to manage billing/payment details (as a Gitpod user - not as the GitPod Stripe admin) ?

**Comment by jldec** [comment]
@easyCZ do you know why the feature flag feels slow ?

**Comment by jankeromnes** [comment]
> Works for me as well

Thanks! I'm glad this works perfectly for everyone. 😊

> Is there a way to navigate manually to the Stripe portal to manage billing/payment details (as a Gitpod user - not as the GitPod Stripe admin) ?

I don't think so, but I'm planning to implement the Stripe portal in the next PR once this is merged.

**Comment by easyCZ** [comment]
> @easyCZ do you know why the feature flag feels slow ?

To me it looks like react is not being told the values have changed so it doesn't propagate as soon as it should. I'm no expert on react either so not entirely sure where this delay lies.
The request for feature flags is made early and completes fast so it should be available at that point for rendering. I'll try to debug a bit when I have time but others more versed in React can probs debug this faster :)

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1860]: The service implementation doesn't catch errors, and neither does this handler. This means that 3rd party errors (from a library) would be propagated to the client. Could you please throw non-detailed error to the client and log the underlying error such that we don't accidentally leak details?

The same also goes for `getTeamStripeCustomerId` and `subscribeTeamToStripe`

Related thread: https://github.com/gitpod-io/gitpod/pull/10378/files#r887842876

**Review by easyCZ — CHANGES_REQUESTED** [review]

**Comment by geropl** [comment]
> If you refresh the dashboard, the Usage-Based feature flag takes a while to be re-enabled (so, if you reload, the Usage-Based UI will be hidden for a few seconds, but otherwise still work well)

This is not the case for me, works rather snappy now. :+1:

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1835]: If feature flags don't work as expected in the `server`, I'd vote for not using them right now (a client-side feature flag is enough), and open a separate issue about fixing the server-side feature flags so that we can use them.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1860]: I left this as is, because forwarding errors makes debugging a bit easier while this is still behind a feature flag (and we have the same problem in many other server API methods, so a more holistic fix would be a good idea).

However, it's pretty easy to add a try/catch to the Stripe methods I've added, so I'll do that now, hoping that we can then merge this and make progress.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1835]: @jankeromnes How about adding a pre-pr to make it work on server properly and then using it here? I'm happy to pair if that makes it easier. Boy/girl scout rules apply.

**Review by geropl — APPROVED** [review]
LGTM, tested and works.

I leave it up to you @jankeromnes to address the ["error leak"](https://github.com/gitpod-io/gitpod/pull/10378/files#r891990649) in a follow-up.

**Comment by jankeromnes** [comment]
Adding a hold for the try/catches:

/hold

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1835]: @easyCZ That's the dashboard implementation, right?

The server-side implementation seems to be here: https://github.com/gitpod-io/gitpod/blob/main/components/server/src/experiments.ts

Comment (easyCZ) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1835]: Interesting, wasn't aware this landed (and my IDE resolved the method to the dashboard one). This must be stray from another PR. It also lives under the root of server/src which it should not.

This will in fact work for server. However, there's a second problem with it as it hard-codes a single Client ID which is for the preview/staging environment, but not production.

Let's land as is, and fix it up in a follow-up PR. We'll need to move the file, but also inject the Client ID based on environment we're running in. The current version would also break self-hosted environments as it reaches out to an external source.

**Comment by easyCZ** [comment]
@jankeromnes Let me know when API layer error handling is in and I'll re-review.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts]: Resolved in [[server] Sanitize Stripe error messages sent by server](https://github.com/gitpod-io/gitpod/pull/10378/commits/d31c9f0af53fdd4f088fd028f3506b1905be616b)

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1865]: Odd we don't have a 500 status code in our list of `errors.ts`

**Review by easyCZ — APPROVED** [review]

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1875]: Also, IMO you should return a 404 error when customer.id does not exist. Returning undefined prevents you classifying your errors client side and actually having reasonable metrics on what worked and didn't. This goes for `setupIntent` also.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1865]: Agree. We hard-code 500 in various places. Should probably fix that 👍 (in a follow-up 😅)

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1875]: Hmm, interesting point. I agree on the 404 semantics. However, returning `undefined` currently means that the customer doesn't exist yet, and is handled that way in the client (different from "the Stripe API failed"). If we switch this around to a 404, we'd need to update the client as well to handle 404 as a valid case (i.e. "the team is not yet a Stripe customer").

Comment (easyCZ) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1875]: I think that's correct. If it doesn't exist, then it wasn't found and we should create one (if needed). In general, we should try to avoid OK responses with No data as much as possible, especially when we know why there is no data - in this case it doesn't exist in stripe.

Comment (easyCZ) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1875]: (Follow-up PR/issue is fine)

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1835]: Follow-up issue: https://github.com/gitpod-io/gitpod/issues/10524

**Pull request merged**
